# I Solving the Sum of Two Digits Programming Challenge

## C++

In [5]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Overwriting APlusB.cpp


In [7]:
!g++ APlusB.cpp -o APlusB
!./APlusB < input1.txt

11

### Test input output files

In [8]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [9]:
%%writefile output1.txt
12

Writing output1.txt


In [10]:
%%writefile input2.txt
-5 9

Overwriting input2.txt


In [11]:
%%writefile output2.txt
4

Writing output2.txt


In [12]:
!./APlusB < input1.txt

12

In [13]:
!./APlusB < input2.txt

4

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [14]:
!grep -xvFf output1.txt output1.txt 

In [15]:
!grep -xvFf output1.txt output2.txt 

4


In [16]:
!grep -xvFf output2.txt output1.txt 

12


In [17]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [18]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [19]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [26]:
!python sum.py < input1.txt

12


In [27]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [28]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [29]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product.cpp


In [35]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!./max_pairwise_product < input1.txt

0


### Python Tests 

Test 1  generate a tests file with the maximun n 20 numbres and measure time

In [3]:
import random
from os import remove

try:
	remove("input4.txt")
except OSError:
	pass

n = 20
f = open("input4.txt", "a")
f.write(str(n)+"\n")
for i in range(n):
	r1 = random.randint(0, 200000)
	f.write(str(r1)+"\n")
f.close()

Test 2  generate a tests file with the maximun n 200000 numbres and measure time

In [9]:
import random
from os import remove

try:
	remove("input3.txt")
except OSError:
	pass

n = 200000-2
f = open("input3.txt", "a")
f.write(str(n+2)+"\n")
f.write("200000\n")
for i in range(n):
	r1 = random.randint(0, 200000)
	f.write(str(r1)+"\n")
f.write("200000")
f.close()

### Test 1: Small numbers

In [53]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [54]:
!./max_pairwise_product < input1.txt

90


### Test 2: largest input numbers 200000 200000 in the input

In [75]:
%%writefile input2.txt
6
1 3 200000 9 200000 23

Overwriting input2.txt


In [56]:
!./max_pairwise_product < input2.txt

1345294336


Changing the variables to long long

In [57]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long.cpp


In [58]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt
!./max_pairwise_product_long < input2.txt

90
40000000000


### Test 3  stress tests 20  times  random n in [2, 20] 

In [59]:
%%writefile Test2Write.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input3.txt");

 for (int i=1; i < 20; i++){
     MyFile << i <<" ";
 }
  
  // Close the file
  MyFile.close();
}

Overwriting Test2Write.cpp


### Test 4  time test  measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end) 

In [60]:
%%writefile Test2Write.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input4.txt");

 MyFile << 200000 <<"\n";  // n 
 for (int i=1; i < 200001; i++){
     MyFile << i <<" ";
 }
  
  // Close the file
  MyFile.close();
}

Overwriting Test2Write.cpp


In [61]:
!g++ Test2Write.cpp -o Test2Write
!./Test2Write

Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [47]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input3.txt

zsh:1: no such file or directory: input3.txt
./max_pairwise_product_long < input3.txt  0.00s user 0.00s system 83% cpu 0.000 total


##### Calculate  the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line 

```cpp
 int main() {                         //N inst exec by line executed  
   
    int n;                            //1                      
    std::cin >> n;                    //1
                                              
    std::vector<int> numbers(n);      //n    init vector
    .
    .
 }

```

In [62]:
%%writefile test_init_vectors.cpp
#include <iostream>
#include <vector>
#include <algorithm>
int main() {                                                               
    int n = 10;                                                                                                                            //1
    std::vector<int> numbers(n);                                            
    for (int i = 0; i < n; ++i) {                                           
        std::cout << numbers[i];                                             
    }

    std::cout << "\n";                       
    return 0;                                                              
}  

Writing test_init_vectors.cpp


In [63]:
!g++ test_init_vectors.cpp -o test_init_vectors
!./test_init_vectors

0000000000


In [69]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed 
    long long max_product = 0LL;                                             //1                                          
	  long long aux_product = 0LL;                                             //1 
    int n = numbers.size();                                                  //1 
                                                                             
    for (int first = 0; first < n; ++first) {                                //1+2n  
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n 
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            if (max_product < aux_product)  {                                //(n-1)n/2 
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
             }
        }
    }
                                                                             // -------------- 
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                              //N inst exec by line executed 
    int n;                                                //1  
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // -------------- 
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]


Overwriting max_pairwise_product_long.cpp


Count executed intructions in code and verify with inputs 1, 2 and 3

In [70]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed 
    
    long long m = 0LL;                                                       //1
    
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1 
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                                //1+2n  
        execints += 2;

        long long m = 0LL;

        execints++;                                    
        for (int second = first + 1; second < n; ++second) {                 //n+ (n-1)n 
            execints += 2;                                                

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            execints++; 

            execints++; 
            if ( max_product < aux_product)  {                               //(n-1)n/2 
                       
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
                execints++; 
                m++;
             }
        }
        if ( (first <= 10) || (n- first <= 10)) std::cout << m << "\n";
    }
                                                                            // -------------- 
    return max_product;                                                     // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;

    int n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }
    
    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";                       
    return 0;                                                // -------------- 
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Overwriting max_pairwise_product_long_execints.cpp


In [72]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt
!./max_pairwise_product_long_execints < input3.txt

9
5
4
3
2
2
2
2
1
0
289
2
1
2
0
0
0
116
10
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
80002600037


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 9 + 7n + [2(n-1)n, 2.5 (n-1)n]$$

It depends on the size of the lists and  the order of the list.

For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 2(9)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 289 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + 2(5)(6)  +  2 + 1 + 2 + 0 + 0 + 0 = 116 $$


For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$
\begin{array}{ll}
T &= 9 + 7(200000) + (2)(199999)(200000)  + (199999 +
100000 + 66667 +50000 \cdots 2 + 2 + 1)\\
&= 80003672077  \\
  &< 9 + 7(200000) + (2.5)(199999)(200000) \\
  & = 100000900009
\end{array}
  $$

Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

128/80003672077 


1.5999265618308827e-09

In [ ]:
# Estimated number of instructions by second

1/(128/80003672077) 

625028688.1015625

### Efficient solution

Add in each line a coment with the number of intstrcution that are executed in that line

In [5]:
%%writefile main.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <ctime>
#include <bits/stdc++.h>

using namespace std;

long long maxPairwiseProduct(const vector<unsigned long long int>& numbers) {
  long long max_product = 0LL;
  long long aux_product = 0LL;
  int n = numbers.size();

  for (int first = 0; first < n; ++first) {
    for (int second = first + 1; second < n; ++second) {
      aux_product =  (long long) numbers[first] * numbers[second];
      if ( max_product < aux_product)  {
        max_product = aux_product;
      }
    }
  }
  return max_product;
}

long long maxPairwiseProductList(vector<unsigned long long int>& numbers){
  sort(numbers.begin(), numbers.end());
  return numbers[numbers.size()-1]*numbers[numbers.size()-2];
}

long long maxPairwiseProductFast(vector<unsigned long long int>& numbers){  
  long long n = numbers.size();
  int index1 = 0;
  int index2 = 0;
    
  for(int i= 1; i < n; i++){
    if(numbers[i] > numbers[index1]){
      index1 = i;
    }
  }
    
  for(int i= 1; i < n; i++){
    if(i!=index1 && numbers[i]>numbers[index2]){
      index2 = i;
    }
  } 
    
  return numbers[index1]*numbers[index2];
}

int main() {
  int n;
  cin >> n;
  unsigned long long int maxNumber = 0LL;
  unsigned long long int maxSecondNumber = 0LL;
  unsigned long long int aux = 0LL;
  vector <unsigned long long int> numbers (n);

  auto start = chrono::high_resolution_clock::now();

  for (int i = 0; i < n; ++i) {
    cin >> numbers[i];
  }
  
  auto stop = chrono::high_resolution_clock::now();
  auto ms_int = chrono::duration_cast<chrono::milliseconds>(stop - start);
  auto reading_time = ms_int.count();
  start = chrono::high_resolution_clock::now();

  cout<<endl;
  cout<<"Start Test..."<<endl;
  cout << "Result using Fast method: " << maxPairwiseProductFast(numbers) << endl;
  stop = chrono::high_resolution_clock::now();
  ms_int = chrono::duration_cast<chrono::milliseconds>(stop - start);
  cout << "Execution time: " << reading_time + ms_int.count() << " ms" << endl;

  cout << endl;
  
  start = chrono::high_resolution_clock::now();
  cout << "Result using List method: " << maxPairwiseProductList(numbers) << endl;
  stop = chrono::high_resolution_clock::now();
  ms_int = chrono::duration_cast<chrono::milliseconds>(stop - start);
  cout << "Execution time: " << reading_time + ms_int.count() << " ms" << endl;

  cout << endl;

  start = chrono::high_resolution_clock::now();
  cout << "Result using naive method: " << maxPairwiseProduct(numbers) << endl;
  stop = chrono::high_resolution_clock::now();
  ms_int = chrono::duration_cast<chrono::milliseconds>(stop - start);
  cout << "Execution time: " << reading_time + ms_int.count() << " ms" << endl;
  return 0;
}

Overwriting main.cpp


Test with inputs 1 and 2



In [6]:
!g++ main.cpp -o main
!./main < input1.txt
!./main < input2.txt


Start Test...
Result using Fast method: 90
Execution time: 0 ms

Result using List method: 90
Execution time: 0 ms

Result using naive method: 90
Execution time: 0 ms

Start Test...
Result using Fast method: 40000000000
Execution time: 0 ms

Result using List method: 40000000000
Execution time: 0 ms

Result using naive method: 40000000000
Execution time: 0 ms


Test and time it with input 3

In [11]:
!./main < input3.txt


Start Test...
Result using Fast method: 40000000000
Execution time: 38 ms

Result using List method: 40000000000
Execution time: 83 ms

Result using naive method: 40000000000
Execution time: 72366 ms


20 stress tests

In [10]:
!g++ main.cpp -o main
!./main < input4.txt


Start Test...
Result using Fast method: 36840938000
Execution time: 0 ms

Result using List method: 36840938000
Execution time: 0 ms

Result using naive method: 36840938000
Execution time: 0 ms


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile main.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <ctime>
#include <bits/stdc++.h>

using namespace std;

long long maxPairwiseProductList(vector<unsigned long long int>& numbers){
  sort(numbers.begin(), numbers.end());                                           //nlog(n)
  return numbers[numbers.size()-1]*numbers[numbers.size()-2];                     //1
}


long long maxPairwiseProduct(const vector<unsigned long long int>& numbers) {
  long long max_product = 0LL;                                                   //1
  long long aux_product = 0LL;                                                   //1
  int n = numbers.size();                                                        //1

  for (int first = 0; first < n; ++first) {                                      //n
    for (int second = first + 1; second < n; ++second) {                         //n
      aux_product =  (long long) numbers[first] * numbers[second];               //1
      if ( max_product < aux_product)  {
        max_product = aux_product;                                               //1
      }
    }
  }
  return max_product;
}

long long maxPairwiseProductFast(vector<unsigned long long int>& numbers){  
  long long n = numbers.size();                                                   //1
  int index1 = 0;                                                                 //1
  int index2 = 0;                                                                 //1
    
  for(int i= 1; i < n; i++){                                                      //n
    if(numbers[i] > numbers[index1]){                                             
      index1 = i;
    }
  }
    
  for(int i= 1; i < n; i++){                                                       //n
    if(i!=index1 && numbers[i]>numbers[index2]){
      index2 = i;
    }
  } 
    
  return numbers[index1]*numbers[index2];                                         //1
}

int main() {
  int n;                                                                           //1
  cin >> n;                                                                        //1
  unsigned long long int maxNumber = 0LL;                                          //1
  unsigned long long int maxSecondNumber = 0LL;                                    //1
  unsigned long long int aux = 0LL;                                                //1
  vector <unsigned long long int> numbers (n);                                     //n

  for (int i = 0; i < n; ++i) {
    cin >> numbers[i];                                                             //n
  }
  
  cout << "maxPairwiseProductList method: " << maxPairwiseProductList(numbers) << endl; //nlog(n)
  cout << "maxPairwiseProduct method: " << maxPairwiseProduct(numbers) << endl; //n^2
  cout << "maxPairwiseProductFast method: " << maxPairwiseProductFast(numbers) << endl; //2n
  
  return 0;
}

Calculate estimated time by instruction and number of instructions by second

maxPairwiseProductList: 5+n+n+nlog(n) = nlog(n)

maxPairwiseProduct: 5+n+n+n+3+n*(n+2)+1 = n^2

maxPairwiseProductFast method: 5+n+n+n+3+n+n+1 = n

### Super efficient solution

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line 

In [16]:
%%writefile main.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <ctime>
#include <bits/stdc++.h>

using namespace std;

int main() {
  int n;
  cin >> n;
  unsigned long long int maxNumber = 0LL;
  unsigned long long int maxSecondNumber = 0LL;
  unsigned long long int aux = 0LL;
  auto start = chrono::high_resolution_clock::now();

  for (int i = 0; i < n; ++i) {
    cin >> aux;
    if(maxNumber < aux){
      maxSecondNumber = maxNumber;
      maxNumber = aux;
    } else if(maxSecondNumber < aux){
      maxSecondNumber = aux;
    }
  }

  unsigned long long int maxProductUltraFast = maxNumber * maxSecondNumber;
  
  auto stop = chrono::high_resolution_clock::now();
  auto ms_int = chrono::duration_cast<chrono::milliseconds>(stop - start);
  auto reading_time = ms_int.count();

  cout << "Result using Ultra Fast method: " << maxProductUltraFast << endl;
  cout << "Execution time: " << ms_int.count() << " ms" << endl;

  return 0;
}

Overwriting main.cpp


Test with inputs 1 and 2


 

In [17]:
!g++ main.cpp -o main
!./main < input1.txt
!./main < input2.txt

Result using Ultra Fast method: 90
Execution time: 0 ms
Result using Ultra Fast method: 40000000000
Execution time: 0 ms


Test and time it with input 3

In [18]:
!./main < input3.txt

Result using Ultra Fast method: 40000000000
Execution time: 33 ms


20 stress tests 

In [19]:
!./main < input4.txt

Result using Ultra Fast method: 36840938000
Execution time: 0 ms


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile main.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <ctime>
#include <bits/stdc++.h>

using namespace std;

int main() {
  int n;                                                    //1
  cin >> n;                                                 //1
  unsigned long long int maxNumber = 0LL;                   //1
  unsigned long long int maxSecondNumber = 0LL;             //1
  unsigned long long int aux = 0LL;                         //1

  for (int i = 0; i < n; ++i) {                             //n
    cin >> aux;                                             //1
    if(maxNumber < aux){                                    //1
      maxSecondNumber = maxNumber;                          //1
      maxNumber = aux;                                      //1
    } else if(maxSecondNumber < aux){                       //1
      maxSecondNumber = aux;                                //1
    }
  }

  unsigned long long int maxProductUltraFast = maxNumber * maxSecondNumber;                          //1
  cout << "Result using Ultra Fast method: " << maxProductUltraFast << endl;                         //1
  return 0;
}

Calculate estimated time by instruction and number of instructions by second

maxProductUltraFast: 5+6n+2 = n

## Python

Efficient cython file reading, string parsing, and array building

https://stackoverflow.com/questions/37219875/efficient-cython-file-reading-string-parsing-and-array-building



File PreProcessing With C and Cython

https://github.com/EveryTimeIWill18/Cython_Repo/blob/master/FastFileProcessingWithCython.ipynb



In [21]:
%%writefile main.py
import time

def maxPairwiseProduct(numbers):
    max_product = 0
    aux_product = 0
    n = len(numbers)
    for first in range(n):
        for second in range(1,n):
            aux_product = numbers[first] * numbers[second]
            if(max_product < aux_product):
                max_product = aux_product
    return max_product

def maxPairwiseProductList(numbers):
	numbers.sort()
	return numbers[-1]*numbers[-2]

def maxPairwiseProductFast(numbers):
    n = len(numbers)
    index1 = 0
    for i in range(n):
        if(numbers[i] > numbers[index1]):
            index1 = i
    
    index2 = 0
    for i in range(n):
        if(numbers[i] != numbers[index1] and numbers[i] > numbers[index2]):
            index2 = i
    return numbers[index1]*numbers[index2]

n = input()
maxNumber = 0
maxSecondNumber = 0
aux = 0
numbers = []

start = time.time()

for i in range(int(n)):
    numbers.append(int(input()))
    if(maxNumber < int(numbers[i])):
        maxSecondNumber = maxNumber
        maxNumber = numbers[i]
    elif(maxSecondNumber < numbers[i]):
        maxSecondNumber = numbers[i]

maxProductUltraFast = maxNumber * maxSecondNumber

stop = time.time()

ms_int = stop-start

reading_time = ms_int

print("Result using Ultra Fast method:",maxProductUltraFast)
print("Execution time:",str(ms_int),"s")
print()

start = time.time()
print("Result using Product Fast method:",maxPairwiseProductFast(numbers))
stop = time.time()
ms_int = stop-start
print("Execution time:",str(float(reading_time + ms_int)),"s")
print()

start = time.time()
print("Result using List method:",maxPairwiseProductList(numbers))
stop = time.time()
ms_int = stop-start
print("Execution time:",str(float(reading_time + ms_int)),"s")
print()

start = time.time()
print("Result using naive method:",maxPairwiseProduct(numbers))
stop = time.time()
ms_int = stop-start
print("Execution time:",str(float(reading_time + ms_int)),"s")

Writing main.py


In [28]:
!timeout 1m python3 main.py < input3.txt

Result using Ultra Fast method: 40000000000
Execution time: 0.16312766075134277 s

Result using Product Fast method: 40000000000
Execution time: 0.18450021743774414 s

Result using List method: 40000000000
Execution time: 0.19104409217834473 s



In [27]:
!timeout 1m python3 main.py < input4.txt

Result using Ultra Fast method: 36840938000
Execution time: 2.5510787963867188e-05 s

Result using Product Fast method: 36840938000
Execution time: 3.266334533691406e-05 s

Result using List method: 36840938000
Execution time: 2.9325485229492188e-05 s

Result using naive method: 36902410000
Execution time: 5.4836273193359375e-05 s


### Cpython read

In [14]:
!pip3 install cython

Defaulting to user installation because normal site-packages is not writeable


In [15]:
%load_ext Cython

The Cython extension is already loaded. To reload it, use:
  %reload_ext Cython


In [25]:
%%cython

import numpy as np
import cython
import time

from libc.stdio cimport *
 
cdef extern from "stdio.h":
    #FILE * fopen ( const char * filename, const char * mode )
    FILE *fopen(const char *, const char *)
    #int fclose ( FILE * stream )
    int fclose(FILE *)
    #ssize_t getline(char **lineptr, size_t *n, FILE *stream);
    ssize_t getline(char **, size_t *, FILE *)
 
def read_file(filename):
    filename_byte_string = filename.encode("UTF-8")
    cdef char* fname = filename_byte_string
    cdef int maxNumber = 0
    cdef int maxSecondNumber = 0
    cdef int aux = 0
    cdef FILE* cfile
    cfile = fopen(fname, "rb")
    if cfile == NULL:
        raise FileNotFoundError(2, "No such file or directory: '%s'" % filename)
 
    cdef char * line = NULL
    cdef size_t l = 0
    cdef ssize_t read
 
    while True:
        read = getline(&line, &l, cfile)
        aux = int(line)
        if(maxNumber < aux):
            maxSecondNumber = maxNumber
            maxNumber = aux
        elif(maxSecondNumber < aux):
            maxSecondNumber = aux           
        if read == -1: break
 
        #yield line
 
    fclose(cfile)
 
    return maxSecondNumber*maxNumber

maxNumber = 0
maxSecondNumber = 0
aux = 0

start = time.time()
numbers = read_file("input3.txt")

stop = time.time()

ms_int = stop-start

reading_time = ms_int

print("Result using Ultra Fast method:",numbers)
print("Execution time:",str(ms_int),"s")

Result using Ultra Fast method: 1345294336
Execution time: 0.024000167846679688 s


# Screenhsoots of the Coursera Submissions in C++ and Python

# Sum Two Digits

<img src="Passed2.png" alt="Drawing"/>

# Maximum Pairwise Product

<img src="Passed1.png" alt="Drawing"/>